fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RelationalExpressionWithPredicate.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * RelationalExpressionWithPredicate.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import com.apple.foundationdb.record.query.predicates.QueryPredicate;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A (relational) expression that has a predicate on it.
+ */
+public interface RelationalExpressionWithPredicate extends RelationalExpression {
+    @Nonnull
+    QueryPredicate getPredicate();
```

Comment:
Yes, `AND` here is encoded using `AndPredicate`.

I agree that it would be pretty nice to have an "implicit top-level AND". In particular, it would let us get rid of the conjunct/non-conjunct distinction in the planner rules.

For the sake of review, I'd like to make this a follow-up PR (which should be fairly short and focussed). I created #939 to track this.